File: C++/ThreeSum.cpp
// Ankit
class Solution {
public:
vector<vector<int>> threeSum(vector<int>& nums) {
    sort(nums.begin(), nums.end());
    vector<vector<int>> r;

    for (int i = 0; i < nums.size() - 2; i++) {
        if (i > 0 && nums[i] == nums[i - 1]) continue;

        int l = i + 1;
        int ri = nums.size() - 1; 

        while (l < ri) {
            int s = nums[i] + nums[l] + nums[ri];

            if (s == 0) {
                r.push_back({nums[i], nums[l], nums[ri]});
                while (l < ri && nums[l] == nums[l + 1]) l++;
                while (l < ri && nums[ri] == nums[ri - 1]) ri--;
                l++;
                ri--;
            } else if (s < 0) {
                l++;
            } else {
                ri--;
            }
        }
    }

    return r;
}
};

File: Rust/ThreeSum.rs
impl Solution {
    pub fn three_sum(mut a: Vec<i32>) -> Vec<Vec<i32>> {
      let mut res = Vec::new();
    a.sort();

    let ra = a.len();
    for i in 0..ra - 2 {
        if i > 0 && a[i] == a[i - 1] {
            continue;   }
        let mut le = i + 1;
        let mut ri = ra - 1;

        while le < ri {
            let s = a[i] + a[le] + a[ri];
            if s == 0 {
                res.push(vec![a[i], a[le], a[ri]]);
                le += 1;
                ri -= 1;

                while le < ri && a[le] == a[le - 1] {
                    le += 1;
                }
                while le < ri && a[ri] == a[ri + 1] {
                    ri -= 1;
                }
            } else if s < 0 {
                le += 1;
            } else {
                ri -= 1;
            }
        }
    }
    // println!("{:?}", res);
    return res;   
    }
}